### DISCUSSION TITLE
JAVA

### DISCUSSION THREAD

**Discussion by seongeun42** [Shortudy-10th/tech-for-developer#5] [오늘의 주제]
자바 질문 리스트

**Comment by sikutisa** [comment]
* 아래 코드의 실행 결과를 예측해보고, 그 이유에 대해 설명해주세요.
```java
import java.io.*;

public class Main {

  public static void main(String[] args) throws IOException {
    String str1 = "string";
    String str2 = "string";
    String str3 = new String("string");
    String str4 = new StringBuilder("string").toString();

    System.out.println(str1 == str2);
    System.out.println(str2 == str3);
    System.out.println(str3 == str4);
    System.out.println(str1.equals(str4));
  }

}
```

**Comment by sikutisa** [comment]
자바소스파일의 컴파일 & 실행과정에 대해 설명해주세요.

**Comment by sikutisa** [comment]
String, StringBuilder, StringBuffer에 대해 설명해주세요.

**Comment by sikutisa** [comment]
String이 Immutable(불변)함으로 인해 얻을 수 있는 이점에는 어떤 것이 있을까요?

**Comment by sikutisa** [comment]
자바의 자료형에 대해 아는 데로 설명해주세요.

**Comment by Jihn0118** [comment]
추상 클래스와 인터페이스의 차이점에 대해서 말씀해주세요.

**Comment by Jihn0118** [comment]
자바 컬렉션에 대해서 설명해주세요.

**Comment by Jihn0118** [comment]
자바에서 배열과 리스트의 차이점에 대해서 말씀해주세요.

**Comment by Jihn0118** [comment]
객체 지향 프로그래밍의 4가지 특징에 대해서 말씀해주세요.

**Comment by Jihn0118** [comment]
자바에서 final 키워드를 변수, 메서드, 클래스에 붙여서 사용할 때 각각 어떤 역할을 하는 지 말씀해주세요.

**Comment by zini9188** [comment]
오버라이딩과 오버로딩에 대해 설명해주세요

**Comment by zini9188** [comment]
static 키워드에 대해 설명하고 장점과 단점에 대해 말해주세요

**Comment by zini9188** [comment]
자바에서 스레드를 사용하기 위한 두 가지 방법에 대해 설명하고 차이점을 알려주세요

**Comment by zini9188** [comment]
try-with-resources와 try-catch-finally에 대해 설명하고 차이점을 알려주세요

**Comment by zini9188** [comment]
자바에서 동시성 문제를 해결하기 위해 어떤 방법을 사용하는지 알려주세요

**Comment by kjc6735** [comment]
checked exception과  unchecked exception 차이를 말해주세요

**Comment by kjc6735** [comment]
동등성과 동일성의 차이와 자바에서는 그 두개를 어떻게 비교하나요?

**Comment by kjc6735** [comment]
Wrapper class가 무엇이며, boxing과 unboxing은 무엇인지 설명해주세요.

**Comment by kjc6735** [comment]
싱글톤 패턴에 대한 설명을 해주세요 (개념, 장단점)

**Comment by kjc6735** [comment]
call by value / call by reference 에 대해 설명해주세요

**Comment by watty-exclamationMark** [comment]
JAVA에는 접근 제어자가 4가지 있습니다.
각각의 이름을 말씀해 주시고, 각각 제어자의 차이점을 설명해 주세요.

**Comment by watty-exclamationMark** [comment]
Error와 Exception의 차이점은 무엇일까요?

**Comment by watty-exclamationMark** [comment]
소프트웨어에서 NullPointer 문제가 많다고 합니다.
null과 관련된 오류가 흔히 발생하고 있다는 것을 알 수 있는데요,
JAVA에서 null을 최대한 안전하게 다루는 방법에 대해 설명해 주세요.

**Comment by watty-exclamationMark** [comment]
JAVA의 메모리 영역에 대해 설명해 주세요.

**Comment by watty-exclamationMark** [comment]
JAVA에서 변수는 멤버 변수와 로컬 변수가 있습니다.
둘의 차이점은 무엇인가요?

**Comment by seongeun42** [comment]
JVM이 무엇이고, 장단점은 무엇인지 설명해주세요.

**Comment by seongeun42** [comment]
가비지 컬렉션이 무엇이고, 장단점은 무엇인지 설명해주세요.

**Comment by seongeun42** [comment]
제네릭이 무엇이고, 왜 사용하는지 설명해주세요.

**Comment by seongeun42** [comment]
Java의 main 메서드가 static으로 선언되는 이유에 대해 설명해주세요.

**Comment by seongeun42** [comment]
아래 코드의 실행 결과를 예측하고, 그 이유에 대해 간단히 설명해주세요.

```java
public class Main {

    public static void main(String[] args) {

        Integer a = Integer.valueOf("10");
        Integer b = 10;
        Integer c = new Integer(10);
        Integer d = Integer.parseInt("10");

        System.out.println(a == b);
        System.out.println(a == c);
        System.out.println(a.equals(c));
        System.out.println(a == d);
        System.out.println(b == d);

    }

}
```

**Comment by Hohyun-Kim** [comment]
자바에서 lambda 함수는 언제 어떻게 사용할 수 있나요?

**Comment by Hohyun-Kim** [comment]
Encapsulation이 뭐고 왜 하는 건가요?

**Comment by Hohyun-Kim** [comment]
입출력시 Stream 과 Reader/Writer의 차이가 무엇인가요?

**Comment by Hohyun-Kim** [comment]
객체 직렬화는 왜 하는건가요?

**Comment by Hohyun-Kim** [comment]
이벤트 리스너가 이벤트를 파악하는 과정을 설명해주세요

**Comment by aeroej** [comment]
JSP의 역할은?

**Comment by aeroej** [comment]
추상 클래스와 일반 클래스의 차이점은?

**Comment by aeroej** [comment]
Autoboxing에 대해 아는 만큼 답변해보세요.

**Comment by aeroej** [comment]
생성자에 대해 아는 만큼 답변해보세요.

**Comment by aeroej** [comment]
싱글톤 패턴이 뭔가요?

**Comment by LireEruel** [comment]
원시타입과 참조타입의 차이에 대해 설명해주세요.

**Comment by LireEruel** [comment]
직렬화와 역직렬화에 대해서 설명해주세요.

**Comment by LireEruel** [comment]
자바가 순수 객체지향언어가 아닌 이유에 대해 설명해주세요

**Comment by LireEruel** [comment]
자바 컴파일러가 컴파일한 결과물은 어떤 파일이죠?

**Comment by LireEruel** [comment]
JVM Runtime data area 구조에 대해 이야기해보세요.
